adaptive_hci/controllers.py

```diff
@@ -1,18 +1,81 @@
+import re
 import lightning.pytorch as pl
 import torch
-from torch.functional import F
+import torch.nn as nn
+from torch.nn import functional as F
 from torchmetrics import ExactMatch, F1Score, Accuracy
 from vit_pytorch import ViT
 
 
 class EMGViT(ViT):
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
+    def __init__(self, config, *args, **kwargs):
+        super().__init__(image_size=config.window_size,
+            patch_size=config.base_model.patch_size,
+            num_classes=config.num_classes,
+            dim=config.base_model.dim,
+            depth=config.base_model.depth,
+            heads=config.base_model.heads,
+            mlp_dim=config.base_model.mlp_dim,
+            dropout=config.base_model.dropout,
+            emb_dropout=config.base_model.emb_dropout,
+            channels=config.base_model.channels,
+            *args, **kwargs)
 
     def __call__(self, x: torch.Tensor):
         x.unsqueeze_(axis=1)
         return self.forward(x)
 
+    def freeze_layers(self, n_frozen_layers: int) -> None:
+        # reset grad
+        for param in self.parameters():
+            param.requires_grad = True
+
+        # freeze desired ones
+        if n_frozen_layers >= 1:
+            for param in self.to_patch_embedding.parameters():
+                param.requires_grad = False
+            for param in self.dropout.parameters():
+                param.requires_grad = False
+
+        if n_frozen_layers >= 2:
+            for layer_idx in range(min((n_frozen_layers - 1), len(self.transformer.layers))):
+                for param in self.transformer.layers[layer_idx].parameters():
+                    param.requires_grad = False
+
+
+class EMGLSTM(nn.Module):
+    def __init__(self, config, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        # TODO feature encoder?
```

Comment:
@manuel-delverme do you think we should have some sort of EMG encoder before the LSTM?